### PR TITLE
Add practice exercise Zebra Puzzle

### PR DESCRIPTION
**Pull request by rajatnai49**
issue number: #779

### CODE REVIEW

File: exercises/practice/zebra-puzzle/.meta/config.json
Comment (vaeng) [inline exercises/practice/zebra-puzzle/.meta/config.json:6]: I have to redo the config, until then please manually correct this part to:

```suggestion
      ".meta/zebra_puzzle.cpp",
      ".meta/zebra_puzzle.h"
```

File: exercises/practice/zebra-puzzle/.meta/example.cpp
Comment (vaeng) [inline exercises/practice/zebra-puzzle/.meta/example.cpp:1]: This is the most difficult part of the puzzle. 
I would prefer this solution to solve the riddle programmatically, rather than having the answer hardcoded. Do you think you can do that?

**Review by vaeng — CHANGES_REQUESTED** [review]
Please add a programmatic way of solving this puzzle.

You can define enums, functions and other more complicated testing scenarios to achieve this goal.

File: exercises/practice/zebra-puzzle/.meta/config.json
Comment (siebenschlaefer) [reply exercises/practice/zebra-puzzle/.meta/config.json:6]: > I have to redo the config, until then please manually correct this part to:

Are you sure? I think these two entries for the initial stub are correct, the example solution (`example.h`, `example.cpp`) should be moved to the `.meta/` directory.

Comment (vaeng) [reply exercises/practice/zebra-puzzle/.meta/config.json:6]: I think our config.json is not correct, so the files and the part above are not generated according to the rest of the track:

https://github.com/exercism/cpp/blob/dad723ff78e3054f917469895b52c7fb66217f71/config.json#L30

**Comment by vaeng** [comment]
Thanks for the update!

Do you think we could have the details of the story in the test file? So the result is not always the same and students cannot just answer "Norwegian"?

So that we have a custom test case, where we give another story? I will ask the other maintainers if they did that. What do you think @siebenschlaefer ?

It is possible a bigger rewrite and I think you should get more than the standard 12 reputation for that, if you want to continue.

**Comment by siebenschlaefer** [comment]
Hej @rajatnai49,

I'm a little bit torn. The example solution still contains a lot of "human reasoning", e.g. why the first house is yellow is deduced from facts 6, 15, and 2 and then hard-coded (`houses[0].color = YELLOW;`)  
I would prefer an example solution that does that as little as possible, let your program find the answer.

For example, you could use five nested loops (for the nationality, the color, ...), try all permutations, and skip a particular permutation as soon as possible when it doesn't match the facts. Or you could try to implement a completely different approach, that's up to you.  
If you need inspiration you could look at the solutions of this exercise in other tracks or at [rosettacode](https://rosettacode.org/wiki/Zebra_puzzle).
And if you need help, let me know, either here or on our Discord server.

And two small things:
- Why are `drinks_water()` and `owns_zebra()` member functions of a `class`? Unless that `class` has some state I'd suggest defining them as free functions instead.
- For constants use `const` or `constexpr` variables instead of macros. That's a broadly accepted best practice.

**Comment by rajatnai49** [comment]
Sorry @siebenschlaefer for disappointment😞, and I know that all permutation solution but I think it will be good if we just first remove that and after that try every possible cases for perticular conditions, and it is much efficient than trying all permutation
But, I understood your point that you want code to do every work, so fine I will work
And also I will take note of those practices this time around!!

**Comment by siebenschlaefer** [comment]
> Sorry @siebenschlaefer for disappointment😞

No worries, please! As vaeng [wrote earlier](https://github.com/exercism/cpp/issues/779#issuecomment-1913264259) this is not an easy exercise. And maybe we maintainer should have communicated more clearly what kind of example solutions we're looking for.

And if you need any help, let us know.

**Comment by siebenschlaefer** [comment]
@rajatnai49 Are you still working on this? Is there anything we can do to help you?

**Comment by rajatnai49** [comment]
Hey, @siebenschlaefer sorry for delay but we are organizing the Hackathon in the university so I didn't have much time to work on the issue so I will give you update in next week for sure!!

**Comment by rajatnai49** [comment]
@siebenschlaefer I added the solution as you explained earlier, review that and if any other requirements are there then let me know !

**Comment by siebenschlaefer** [comment]
Much better, this is now an algorithmic solution.

---

But it takes a really long time to run because it tries all possible potential and finds the solution after more than 171 million permutations. On my computer it runs in a a little bit more than one minute. Students wouldn't be able to submit this example solution because it would not complete the tests within the allowed time (I think the test runner has a limit of 20 or 30 seconds.)  
We need to bring that time down because we want an example solution that *could* be successfully submitted.

There are five nested loops (for the `cigarettes`, `drinks`, `pets`, `nationalities`, and `colors`). Currently the innermost loop assigns values to the members of the `houses` and checks whether that's a valid solution.

Instead you could assign assign the values to the houses as soon as possible, i.e. the `.cigarette` in the outer loop, the `.drink` in the first inner loop, the `.pet` in the second inner loop, the `.nationality` in the third inner loop, and the `.color` in the innermost loop. That alone cuts the runtime in half (at least on my computer).  

Instead of checking the complete solution in the innermost loop you could check each of the 14 clauses as soon as possible and `continue` if they don't hold. For example the outermost loop creates permutations of the cigarette brands, the first inner loop creates permutations of the drinks. You could check clauses 8 ("Milk is drunk in the middle house") and 12 ("The Lucky Strike smoker drinks orange juice.") right there, in the first inner loop. That would allow you to skip generating all the permutations of pets, nationalities, and colors when you know that the will never create a valid solution because `houses[2].drink` is not `Milk` or because the Lucky Strike smoker does not drink OJ. If you do that for all clauses (check each clause in the outermost possible loop) you can reduce the runtime drastically, on my computer it runs in under one second.

---

And some smaller nitpicks: 

Four of the five `std::map` instances are not used at all. I'd suggest deleting them. That makes the code easier to read and to maintain.

As far as I can see the `return` statement in line 159 is the only one that is not indented by a multiple of four spaces.

---

And finally a more general question where @vaeng might also want to weigh in:

Currently the tests require two separate and free functions that answer the two questions. But is that really the best structure for this problem. It basically forces the students to write some helper function that solves the problem and gets called by the two "public" functions, and it will make the test runner to solve the puzzle twice (unless the students implement some caching) without a clear benefit.

How about a single function that returns both answers at the same time? That's how the Go track does it.

File: exercises/practice/zebra-puzzle/.meta/config.json
Comment (vaeng) [inline exercises/practice/zebra-puzzle/.meta/config.json:13]:
```suggestion
      ".meta/example.cpp",
      ".meta/example.h"
```

**Review by vaeng — DISMISSED** [review]

**Comment by vaeng** [comment]
> How about a single function that returns both answers at the same time? That's how the Go track does it.

We could also go the class way and generate the puzzle solution on construction and have the answers in variables ... Might be ugly :D

**Comment by siebenschlaefer** [comment]
> We could also go the class way and generate the puzzle solution on construction and have the answers in variables ... Might be ugly :D

Right, that's another option. (although *personally* I'm not a fan)

**Comment by rajatnai49** [comment]
@siebenschlaefer totally understood ☺️, working on that. and both single function or class construction approaches are great because there is a only one particular answer for the zebra puzzle, so solving puzzle two time, like in the python track will not be add any value, So let me know on which path we are going.

**Comment by vaeng** [comment]
> So let me know on which path we are going.

a single function that returns both answers at the same time.

**Comment by rajatnai49** [comment]
@vaeng @siebenschlaefer 
can you please give me the status/review of this PR, anything required?

File: exercises/practice/zebra-puzzle/.meta/example.cpp
Comment (siebenschlaefer) [inline exercises/practice/zebra-puzzle/.meta/example.cpp:16]: If you '#include` the `.h` file you won't need to repeat this definition (see above).

File: exercises/practice/zebra-puzzle/.meta/example.h
Comment (siebenschlaefer) [inline exercises/practice/zebra-puzzle/.meta/example.h:13]: `typedef struct { ... } type_name;` is a C idiom.  
In C++ you can write `struct type_name { ... };`, just like you did with `House` in the `.cpp` file.

File: exercises/practice/zebra-puzzle/.meta/tests.toml
Comment (siebenschlaefer) [inline exercises/practice/zebra-puzzle/.meta/tests.toml:13]: This file should describe which tests from the language-independent [problem specification](https://github.com/exercism/problem-specifications/blob/main/exercises/zebra-puzzle/canonical-data.json) are implemented in this track, the UUID and description should match those from the JSON file.  
The `zebra_puzze_tests.cpp` implements both tests ("resident who drinks water" and "resident who owns zebra") so this TOML file should reflect that.

File: exercises/practice/zebra-puzzle/zebra_puzzle.cpp
Comment (siebenschlaefer) [inline exercises/practice/zebra-puzzle/zebra_puzzle.cpp:5]: While even some C++ programmers like to have this `using namespace std;`  in their `.cpp` files most C++ experts recommend not to do that. I'd suggest removing that line, we don't want to promote bad practices.

File: exercises/practice/zebra-puzzle/zebra_puzzle.cpp
Comment (siebenschlaefer) [inline exercises/practice/zebra-puzzle/zebra_puzzle.cpp:3]: This is not necessary, `<string>` is already included in the `.h` file.

File: exercises/practice/zebra-puzzle/zebra_puzzle.cpp
Comment (siebenschlaefer) [inline exercises/practice/zebra-puzzle/zebra_puzzle.cpp:14]: *\*If\** you want to define that type for our students, do that in the `.h` file where it is required.  
(and without the `typedef` idiom, see above.)

File: exercises/practice/zebra-puzzle/zebra_puzzle.cpp
Comment (siebenschlaefer) [inline exercises/practice/zebra-puzzle/zebra_puzzle.cpp:19]: *If* you want to provide a function stub for the students as a hint which function they must implement, and what it returns, declare it in the `.h` file, like in the other exercises that do this.

**Review by siebenschlaefer — COMMENTED** [review]
This looks good!

There are a few smaller things that I'd like to improve before merging.

The only potential problem that I can see is that the tests use `SECTION` which does not map directly to the two tests in the [problem specification](https://github.com/exercism/problem-specifications/blob/main/exercises/zebra-puzzle/canonical-data.json). The test name "drinks water" and UUID does not reflect that combined nature of the tests. I'd rather have two separate and independent tests.  

@vaeng what do you think?

**Comment by vaeng** [comment]
@rajatnai49 
The exercise is going to be featured in two weeks. Can you cover siebenschläfer's suggestions in the next days, so we can have it online soon?

**Review by siebenschlaefer — APPROVED** [review]
Looks good. Thank you very much!

**Review by vaeng — APPROVED** [review]

**Comment by vaeng** [comment]
This is it. Thaks a lot for all the work in this exercise @siebenschlaefer and especially @rajatnai49!

**Pull request merged**
